### PR TITLE
gh-81057: Move the global Dict-Related Versions to _PyRuntimeState

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
We also move the global func version.

* Issue: gh-81057

### CODE REVIEW

File: Include/internal/pycore_dict_state.h [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+#define DICT_MAX_WATCHERS 8
+
+struct _Py_dict_state {
+#if PyDict_MAXFREELIST > 0
+    /* Dictionary reuse scheme to save calls to malloc and free */
+    PyDictObject *free_list[PyDict_MAXFREELIST];
+    int numfree;
+    PyDictKeysObject *keys_free_list[PyDict_MAXFREELIST];
+    int keys_numfree;
```
Comment (methane) [inline Include/internal/pycore_dict_state.h]: How about reorder members to make struct compact?

```suggestion
    PyDictObject *free_list[PyDict_MAXFREELIST];
    PyDictKeysObject *keys_free_list[PyDict_MAXFREELIST];
    int numfree;
    int keys_numfree;
```

**Review by methane — APPROVED** [review]

File: Include/internal/pycore_dict_state.h (hunk 1 above)
Comment (ericsnowcurrently) [reply Include/internal/pycore_dict_state.h]: I don't mind doing that, but it isn't much related to this PR and I was trying to minimize extra changes.  That said, it is a fairly small and simple change so I'll go ahead and do it.

Comment (ericsnowcurrently) [reply Include/internal/pycore_dict_state.h]: done

**Comment by ericsnowcurrently** [comment]
Thanks for the review, @methane!  For a domain-specific change to a complex component like this, it really helps to get an expert review, even if the change seems correct and straight-forward to me. 😄

**Pull request merged**